### ISSUE TITLE
OutOfMemoryException in EventBasedLRUCache after upgrading to Microsoft.IdentityModel.Tokens 6.10.1

### ISSUE THREAD

**Issue by nemesv** [AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet#1630]
We are experiencing increased Thread counts and OutOfMemoryExceptions after upgrading to System.IdentityModel.Tokens.Jwt version 6.10.1 in our ASP.NET Core application running in Azure.

After rolling back the package to 6.9.0 the Thread counts are normalized:
![image](https://user-images.githubusercontent.com/251330/115038063-75ff0c00-9ecf-11eb-8c2b-ec5a3f61d53f.png)

The relevant part of the stack trace of the OutOfMemoryExceptions reported in AppInsigts:

```
{"assembly":"System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e","method":"System.Threading.Tasks.Task.ScheduleAndStart","level":0,"line":0}
{"assembly":"System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e","method":"System.Threading.Tasks.Task.Start","level":1,"line":0}
{"assembly":"Microsoft.IdentityModel.Tokens, Version=6.10.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35","method":"Microsoft.IdentityModel.Tokens.EventBasedLRUCache`2..ctor","level":2,"line":0}
{"assembly":"Microsoft.IdentityModel.Tokens, Version=6.10.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35","method":"Microsoft.IdentityModel.Tokens.InMemoryCryptoProviderCache..ctor","level":3,"line":0}
{"assembly":"Microsoft.IdentityModel.Tokens, Version=6.10.1.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35","method":"Microsoft.IdentityModel.Tokens.CryptoProviderFactory..ctor","level":4,"line":0}
{"assembly":"redacted, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null","method":"redacted.AzureKeyVaultTokenSignatureService.CreateSignedToken"}

```

We are using the following code to sign JWT tokens using AzureKeyVault:

```c#
public string CreateSignedToken(JwtPayload jwtPayload)
{
    var azureServiceTokenProvider = new AzureServiceTokenProvider();

    var signingCredentials =
        new SigningCredentials(
            new KeyVaultSecurityKey($"https://{configuration["KeyVaultName"]}.vault.azure.net/keys/{configuration["redacted"]}",
                new KeyVaultSecurityKey.AuthenticationCallback(azureServiceTokenProvider
                    .KeyVaultTokenCallback)), SecurityAlgorithms.RsaSha512)
        {
            CryptoProviderFactory = new CryptoProviderFactory()
                { CustomCryptoProvider = new KeyVaultCryptoProvider() }
        };

    var jwtHeader = new JwtHeader(signingCredentials);

    var token = new JwtSecurityToken(jwtHeader, jwtPayload);

    var handler = new JwtSecurityTokenHandler();
    return handler.WriteToken(token);
}
```

Here is the diff our our change which caused the errors start to happen:

```diff
- <PackageReference Include="Microsoft.IdentityModel.KeyVaultExtensions" Version="6.9.0" />
- <PackageReference Include="System.IdentityModel.Tokens.Jwt" Version="6.9.0" />
+ <PackageReference Include="Microsoft.IdentityModel.KeyVaultExtensions" Version="6.10.1" />
+ <PackageReference Include="System.IdentityModel.Tokens.Jwt" Version="6.10.1" />
```

We are on .NET Core 3.1.407.

**Comment by mafurman** [comment]
I believe the issue here is with the way that you are creating the CryptoProviderFactory.

Each CryptoProviderFactory has it's own instance of the EventBasedLRUCache, which means it spins up it's own thread for cache processing purposes. This is causing too many threads to build up in the background as you are creating a new factory/thread each time you sign a token.

Threading aside, creating a new factory each time prevents you from making use of the signature provider caching feature, which improves performance. Can you modify your code to reuse the same CryptoProviderFactory across calls?

**Comment by nemesv** [comment]
We could modify our code to reuse the `CryptoProviderFactory` but we would like to understand the issue better. 

- can you point us to some documentation where it is explained how the `CryptoProviderFactory` should be used correctly?
- should we also keep around the `KeyVaultCryptoProvider` or that one should be created for each request?
- are you plan to change/fix this caching behavior in an upcoming version?

**Comment by nemesv** [comment]
@mafurman do you have any updates on this issue?

**Comment by dmetzgar** [comment]
+1 on this issue. After deploying with release 6.11, the CPU usage on our pods steadily rose over a 1.5 hour period until the node CPU hit 100% and the service became unresponsive. A core dump showed >70 threads stuck in this callstack:

```
System.Private.CoreLib.dll!System.Threading.SemaphoreSlim.WaitUntilCountOrTimeout(int millisecondsTimeout, uint startTime, System.Threading.CancellationToken cancellationToken) Line 462	C#
System.Private.CoreLib.dll!System.Threading.SemaphoreSlim.Wait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 376	C#
System.Collections.Concurrent.dll!System.Collections.Concurrent.BlockingCollection<System.Action>.TryTakeWithNoTimeValidation(out System.Action item, int millisecondsTimeout, System.Threading.CancellationToken cancellationToken, System.Threading.CancellationTokenSource combinedTokenSource)	Unknown
Microsoft.IdentityModel.Tokens.dll!Microsoft.IdentityModel.Tokens.EventBasedLRUCache<string, Microsoft.IdentityModel.Tokens.SignatureProvider>.OnStart()	Unknown
System.Private.CoreLib.dll!System.Threading.Tasks.Task.InnerInvoke() Line 2380	C#
System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) Line 186	C#
System.Private.CoreLib.dll!System.Threading.Tasks.Task.ExecuteWithThreadLocal(ref System.Threading.Tasks.Task currentTaskSlot, System.Threading.Thread threadPoolThread) Line 2316	C#
System.Private.CoreLib.dll!System.Threading.Tasks.ThreadPoolTaskScheduler..cctor.AnonymousMethod__10_0(object s) Line 36	C#
System.Private.CoreLib.dll!System.Threading.ThreadHelper.ThreadStart_Context(object state) Line 48	C#
System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) Line 186	C#
System.Private.CoreLib.dll!System.Threading.ThreadHelper.ThreadStart(object obj) Line 81	C#
```

Upgrading to 6.12.0 fixed the issue.

**Comment by mackie1001** [comment]
I'd just like to add a note here to say we encountered the same issue as @dmetzgar when upgrading from 5.6.0 to 6.11.1. After being under normal workday load for ~6 hours or so we'd see a pretty steady rise in CPU usage until it maxed out. 

Out of interest what tooling did you use to isolate that @dmetzgar ?

6.12 appears to fix it for us but we will continue to monitor.

We were not creating a new CryptoProviderFactory manually but the application uses System.IdentityModel.Tokens.Jwt heavily for both signing and validation of tokens (it's an IdentityServer4 based service). Is anyone able to shed some light on how this came about with normal usage of the library?

**Comment by dmetzgar** [comment]
@mackie1001 To get the stacks, we did a core dump of the container and opened it in Visual Studio. VS is limited on Linux dumps but at least gives a thread view where we can see the stacks on each thread. I later confirmed with LLDB the number of threads.

**Comment by brentschmaltz** [comment]
@dmetzgar @mackie1001 @nemesv thanks for the detail.
Our perf and stress tests didn't pick this up because they don't create a new CryptoProviderFactory, ill add that.
We had a bug in our code where a task was left running, we assumed incorrectly, a single CryptoProviderFactory in play.

It is probably more efficient when you need to support a new algorithm to make use of CryptoProviderFactory.CustomCryptoProvider as we will call into that object first and if you respond true to IsSupportedAlgorithm, we will pass you control.

I will close this out for now, please ping us if you have issues.

**Comment by dannybtsai** [comment]
Note this issue was fixed by #1650.
